packages/compiler/src/core/name-resolver.ts

```diff
@@ -0,0 +1,1316 @@
+ * program.
+ **/
+
+import { Mutable, mutate } from "../utils/misc.js";
+import { createSymbol, createSymbolTable } from "./binder.js";
+import { compilerAssert } from "./diagnostics.js";
+import { typeReferenceToString } from "./helpers/syntax-utils.js";
+import { createDiagnostic } from "./messages.js";
+import { visitChildren } from "./parser.js";
+import { Program } from "./program.js";
+import { getFullyQualifiedSymbolName } from "./type-utils.js";
+import {
+  AliasStatementNode,
+  AugmentDecoratorStatementNode,
+  DecoratorExpressionNode,
+  EnumStatementNode,
+  Expression,
+  IdentifierNode,
+  InterfaceStatementNode,
+  IntersectionExpressionNode,
+  MemberExpressionNode,
+  ModelExpressionNode,
+  ModelPropertyNode,
+  ModelStatementNode,
+  NamespaceStatementNode,
+  Node,
+  NodeFlags,
+  NodeLinks,
+  OperationStatementNode,
+  ProjectionDecoratorReferenceExpressionNode,
+  ProjectionStatementNode,
+  ResolutionResult,
+  ResolutionResultFlags,
+  ScalarStatementNode,
+  Sym,
+  SymbolFlags,
+  SymbolLinks,
+  SymbolTable,
+  SyntaxKind,
+  TemplateParameterDeclarationNode,
+  TypeReferenceNode,
+  TypeSpecScriptNode,
+  UnionStatementNode,
+} from "./types.js";
+
+export interface NameResolver {
+  // TODO: add docs
+  resolveProgram(): void;
+
+  /**
+   * Get the merged symbol or itself if not merged.
+   * This is the case for Namespace which have multiple nodes and symbol but all reference the same merged one.
+   */
+  getMergedSymbol(sym: Sym): Sym;
+
+  /**
+   * Get augmented symbol table.
+   */
+  getAugmentedSymbolTable(table: SymbolTable): Mutable<SymbolTable>;
+  /**
+   * Get node links for the given syntax node.
+   * This returns links to which symbol the node reference if applicable(TypeReference, Identifier nodes)
+   */
+  getNodeLinks(node: Node): NodeLinks;
+
+  /** Get symbol links for the given symbol */
+  getSymbolLinks(symbol: Sym): SymbolLinks;
+
+  /** Get global namespace symbol. */
+  getGlobalNamespaceSymbol(): Sym;
+
+  /** Return augment decorator nodes that are bound to this symbol */
+  getAugmentDecoratorForSym(symbol: Sym): AugmentDecoratorStatementNode[];
+
+  /**
+   * Resolve the member expression using the given symbol as base.
+   * This can be used to follow the name resolution for template instance which are not statically linked.
+   */
+  resolveMemberExpressionForSym(
+    sym: Sym,
+    node: MemberExpressionNode,
+    options?: ResolveTypReferenceOptions,
+  ): ResolutionResult;
+
+  /** Get the meta member by name */
+  resolveMetaMemberByName(sym: Sym, name: string): ResolutionResult;
+
+  /** Resolve the given type reference. This should only need to be called on dynamically created nodes that want to resolve which symbol they reference */
+  resolveTypeReference(
+    node: TypeReferenceNode | IdentifierNode | MemberExpressionNode,
+  ): ResolutionResult;
+
+  readonly intrinsicSymbols: {
+    readonly null: Sym;
+  };
```

Comment:
What do people think of this pattern vs the `getGlobalNamespaceSymbol`

coudl change to 
```
readonly symbols: {
    readonly globalNamespace: Sym;
    readonly null: Sym;
  };
```

Reply:
I don't like getGlobalNamespaceSymbol personally, so I like this.